File: frontend/js/views/percentage_rent.js
// Percentage-rent statement generator — retail-lease overage: base + rate ×
// sales over the (natural or stated) breakpoint, via /calc/percentage-rent.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

export async function renderPercentageRent(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.pctrent.h1.title">// PERCENTAGE RENT</span></h1>
        <p class="muted small" data-i18n="view.pctrent.hint.intro">
            The retail-lease overage calculation. A tenant pays fixed base rent plus a percentage of gross
            sales above a breakpoint — natural (base rent ÷ rate) unless the lease states one. It computes the
            natural breakpoint, the overage rent, total rent, and the occupancy-cost ratio (total rent ÷
            sales). Drafting aid, not legal/accounting advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.pctrent.h2.inputs">Lease terms</h2>
            <form id="pctrent-form" class="inline-form">
                <label><span data-i18n="view.pctrent.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.pctrent.label.landlord">Landlord</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.pctrent.label.tenant">Tenant</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.pctrent.label.property">Premises</span>
                    <input type="text" name="property_label" value="Store 12"></label>
                <label><span data-i18n="view.pctrent.label.base">Annual base rent ($)</span>
                    <input type="number" step="1000" min="0" name="base_rent_usd" value="120000" required></label>
                <label><span data-i18n="view.pctrent.label.rate">Percentage rate (%)</span>
                    <input type="number" step="0.1" min="0" name="rate_pct" value="6" required></label>
                <label><span data-i18n="view.pctrent.label.sales">Gross sales ($)</span>
                    <input type="number" step="10000" min="0" name="gross_sales_usd" value="3000000" required></label>
                <label><span data-i18n="view.pctrent.label.breakpoint">Stated breakpoint ($, 0 = natural)</span>
                    <input type="number" step="10000" min="0" name="stated_breakpoint_usd" value="0"></label>
                <label><span data-i18n="view.pctrent.label.period">Period (months)</span>
                    <input type="number" step="1" min="1" name="period_months" value="12" required></label>
                <label><span data-i18n="view.pctrent.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.pctrent.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.pctrent.ph.statute'))}"></label>
            </form>
        </div>
        <div id="pctrent-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#pctrent-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            property_label: (fd.get('property_label') || '').trim(),
            base_rent_usd: Number(fd.get('base_rent_usd')) || 0,
            rate_pct: Number(fd.get('rate_pct')) || 0,
            gross_sales_usd: Number(fd.get('gross_sales_usd')) || 0,
            stated_breakpoint_usd: Number(fd.get('stated_breakpoint_usd')) || 0,
            period_months: Number(fd.get('period_months')) || 0,
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcPercentageRent(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.pctrent.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const bpLabel = doc.stated_breakpoint_applied ? t('view.pctrent.card.bp_stated') : t('view.pctrent.card.bp_natural');
    const el = mount.querySelector('#pctrent-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.pctrent.card.total">Total rent</div>
                    <div class="value">${money(doc.total_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pctrent.card.overage">Overage rent</div>
                    <div class="value">${money(doc.overage_rent_usd)}</div></div>
                <div class="card"><div class="label">${esc(bpLabel)}</div>
                    <div class="value">${money(doc.breakpoint_used_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pctrent.card.occupancy">Occupancy cost</div>
                    <div class="value">${pct(doc.occupancy_cost_pct)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="pctrent-copy" type="button" data-i18n="view.pctrent.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="pctrent-download" type="button" data-i18n="view.pctrent.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#pctrent-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.pctrent.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.pctrent.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#pctrent-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'percentage-rent.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/percentage_rent.rs
//! Percentage-rent statement — the retail-lease overage calculation. A tenant
//! pays a fixed base rent plus a percentage of gross sales above a breakpoint.
//! The breakpoint is "natural" (base rent ÷ percentage rate) unless the lease
//! states one. This computes the natural breakpoint, the overage rent, the total
//! rent, and the occupancy-cost ratio (total rent ÷ sales) — the natural-
//! breakpoint and occupancy-ratio computations are specific to retail leasing and
//! appear in no other generator. Drafting aid, not legal/accounting advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct PercentageRentInput {
    pub landlord_name: String,
    pub tenant_name: String,
    #[serde(default)]
    pub property_label: String,
    /// Annual base (minimum) rent.
    pub base_rent_usd: f64,
    /// Percentage-rent rate on sales over the breakpoint, percent.
    pub rate_pct: f64,
    /// Gross sales for the period.
    pub gross_sales_usd: f64,
    /// Stated breakpoint (0 = use the natural breakpoint = base ÷ rate).
    #[serde(default)]
    pub stated_breakpoint_usd: f64,
    /// Period length, in months.
    pub period_months: u32,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct PercentageRentStatement {
    pub title: String,
    /// base ÷ rate — sales level at which percentage rent begins.
    pub natural_breakpoint_usd: f64,
    /// The breakpoint used (stated if given, else natural).
    pub breakpoint_used_usd: f64,
    /// True when a stated breakpoint overrode the natural one.
    pub stated_breakpoint_applied: bool,
    /// Sales above the breakpoint.
    pub sales_over_breakpoint_usd: f64,
    /// Percentage rent: rate × sales over breakpoint.
    pub overage_rent_usd: f64,
    /// Base + overage.
    pub total_rent_usd: f64,
    /// Total rent ÷ sales, percent.
    pub occupancy_cost_pct: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &PercentageRentInput) -> PercentageRentStatement {
    let natural_bp = if i.rate_pct > 0.0 {
        cents(i.base_rent_usd / (i.rate_pct / 100.0))
    } else {
        0.0
    };
    let stated_applied = i.stated_breakpoint_usd > 0.0;
    let breakpoint = if stated_applied {
        i.stated_breakpoint_usd
    } else {
        natural_bp
    };

    let over = (i.gross_sales_usd - breakpoint).max(0.0);
    let overage = cents(over * i.rate_pct / 100.0);
    let total_rent = cents(i.base_rent_usd + overage);
    let occupancy = if i.gross_sales_usd > 0.0 {
        cents(total_rent / i.gross_sales_usd * 100.0)
    } else {
        0.0
    };

    let property = if i.property_label.trim().is_empty() {
        "the Premises".to_string()
    } else {
        i.property_label.trim().to_string()
    };

    let bp_desc = if stated_applied {
        format!("the stated breakpoint of {}", money(breakpoint))
    } else {
        format!("the natural breakpoint of {} (base rent {} ÷ {:.2}%)", money(natural_bp), money(i.base_rent_usd), i.rate_pct)
    };

    let calc_body = format!(
        "Percentage rent is {:.2}% of gross sales above {}. Gross sales of {} exceed the breakpoint by {}, producing overage rent of {}. Added to base rent of {}, total rent is {} — an occupancy cost of {:.2}% of sales.",
        i.rate_pct,
        bp_desc,
        money(i.gross_sales_usd),
        money(over),
        money(overage),
        money(i.base_rent_usd),
        money(total_rent),
        occupancy
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("This statement is governed by the lease and the laws of the State of {}.", i.state)
    } else {
        format!("This statement is governed by the lease and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Landlord: {}\nTenant: {}\nPremises: {}\nPeriod: {} month(s)\nStatement date: {}",
                i.landlord_name, i.tenant_name, property, i.period_months, i.date
            ),
        },
        DocClause {
            heading: "1. Base Rent".into(),
            body: format!("The Tenant pays annual base (minimum) rent of {}.", money(i.base_rent_usd)),
        },
        DocClause {
            heading: "2. Percentage Rent".into(),
            body: format!(
                "In addition to base rent, the Tenant pays percentage rent of {:.2}% of gross sales exceeding {}.",
                i.rate_pct, bp_desc
            ),
        },
        DocClause { heading: "3. Calculation".into(), body: calc_body },
        DocClause {
            heading: "4. Payment & Reporting".into(),
            body: "The Tenant shall report gross sales and pay percentage rent within 30 days after the period. Gross sales exclude returns, sales taxes, and inter-store transfers as defined in the lease.".into(),
        },
        DocClause {
            heading: "5. Records & Audit".into(),
            body: "The Tenant shall keep complete records of gross sales for at least three years; the Landlord may audit them on reasonable notice. An understatement over 3% entitles the Landlord to the audit cost.".into(),
        },
        DocClause { heading: "6. Governing Law".into(), body: governing },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}\n\nTenant: ____________________  Date: __________\n{}",
                i.landlord_name, i.tenant_name
            ),
        },
    ];

    PercentageRentStatement {
        title: "Percentage Rent Statement".into(),
        natural_breakpoint_usd: natural_bp,
        breakpoint_used_usd: cents(breakpoint),
        stated_breakpoint_applied: stated_applied,
        sales_over_breakpoint_usd: cents(over),
        overage_rent_usd: overage,
        total_rent_usd: total_rent,
        occupancy_cost_pct: occupancy,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> PercentageRentInput {
        PercentageRentInput {
            landlord_name: "Mall Owners LP".into(),
            tenant_name: "Boutique Tenant LLC".into(),
            property_label: "Store 12".into(),
            base_rent_usd: 120_000.0,
            rate_pct: 6.0,
            gross_sales_usd: 3_000_000.0,
            stated_breakpoint_usd: 0.0,
            period_months: 12,
            date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn natural_breakpoint_and_overage() {
        let d = generate(&base());
        assert!(close(d.natural_breakpoint_usd, 2_000_000.0));
        assert!(close(d.breakpoint_used_usd, 2_000_000.0));
        assert!(!d.stated_breakpoint_applied);
        assert!(close(d.sales_over_breakpoint_usd, 1_000_000.0));
        assert!(close(d.overage_rent_usd, 60_000.0));
        assert!(close(d.total_rent_usd, 180_000.0));
        assert!(close(d.occupancy_cost_pct, 6.0));
    }

    #[test]
    fn stated_breakpoint_overrides_natural() {
        let d = generate(&PercentageRentInput { stated_breakpoint_usd: 2_500_000.0, ..base() });
        assert!(d.stated_breakpoint_applied);
        assert!(close(d.breakpoint_used_usd, 2_500_000.0));
        assert!(close(d.overage_rent_usd, 30_000.0));
        assert!(close(d.total_rent_usd, 150_000.0));
        assert!(close(d.occupancy_cost_pct, 5.0));
        // Natural breakpoint is still reported for reference.
        assert!(close(d.natural_breakpoint_usd, 2_000_000.0));
    }

    #[test]
    fn below_breakpoint_no_overage() {
        let d = generate(&PercentageRentInput { gross_sales_usd: 1_500_000.0, ..base() });
        assert!(close(d.sales_over_breakpoint_usd, 0.0));
        assert!(close(d.overage_rent_usd, 0.0));
        assert!(close(d.total_rent_usd, 120_000.0));
        assert!(close(d.occupancy_cost_pct, 8.0));
    }

    #[test]
    fn occupancy_ratio_equals_total_over_sales() {
        // total_rent 180,000 ÷ sales 3,000,000 × 100 = 6.00%.
        let i = base();
        let d = generate(&i);
        assert!(close(d.occupancy_cost_pct, d.total_rent_usd / i.gross_sales_usd * 100.0));
    }

    #[test]
    fn audit_clause_present() {
        assert!(generate(&base()).clauses.iter().any(|c| c.body.contains("audit")));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&PercentageRentInput { statute_citation: "lease § 3.2".into(), ..base() });
        assert_eq!(d.statutory_citation, "lease § 3.2");
        assert!(d.clauses.iter().any(|c| c.body.contains("lease § 3.2")));
    }
}
